fix: do not crash on unexpected map format in GenericMapTypeHandler

jdrueckert opened the pull request:
### Contains

- detect incorrect map format (object instead of array) early, log expected format
  ```
  "mapName": [
    { "key": "...", "value": "..." }
  ]
  ```
- add null check to avoid crashing if key "key" not found in case of incorrect map format (no map entry for "key")
- add null check to avoid crashing if key "value" not found in case of incorrect map format (no map entry for "value")

### How to test

Can be tested in combination with https://github.com/Terasology/PlantPack/pull/20

### Outstanding before merging

- [ ] judge whether refactoring actually improves readability
- [x] check that refactoring doesn't break anything 
  - [x] discuss the return value in case of mismatched key / value handler

jdrueckert commented on subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/coreTypes/GenericMapTypeHandler.java:
```diff
@@ -61,26 +61,45 @@ private PersistedData serializeEntry(Map.Entry<K, V> entry, PersistedDataSeriali
+            final Optional<K> key = keyHandler.deserialize(rawKey);
+            if (key.isEmpty()) {
+                logger.warn("Could not deserialize key '{}' as '{}'", rawKey, keyHandler.getClass().getSimpleName());
+                return Optional.empty();
```
@keturn I saw the test cases in `GenericMapTypeHandlerTest.java` and wanted to ask, why it's desirable to have an empty map returned in case of a mismatched key / value handler... My current assumption is that we want that because failing to deserialize the `ChangingBlocks` component should not result in the whole prefab deserialization failing (although I'm not even sure that would be the case when returning `Optional.empty()` instead :thinking: 

Can you confirm this assumption or otherwise explain the reason for expecting an empty map in this case?

My intention with using `Optional.empty()` here is to indicate that the serialization failed and to not return a half-way deserialized map as might happen for instance in case the prefab holds:
```
myMap: [
  { "key": "myKey", "value": "myValue" },
  { "key2": "myKey2", "value2": "myValue2" }
]
```
With returning `Optional.of(result)`, I would expect the first map entry to be put into the result map and the deserialization failing for the second entry, resulting in a half-way deserialized map which might lead to weird and potentially hard to debug behavior in game.
Taking the growing plant feature, for example, the plant would never grow to the second stage as the second stage never ended up in game because its deserialization failed.

Please correct me if I got any of the type handling logic etc. wrong leading to incorrect assumptions/expectations :sweat_smile:

keturn replied:
It looks like I wrote those test cases after @skaldarnar wrote the implementation, so we should ask that guy too.

I'm struggling with how to answer these questions because I think you raise some good points, and—

- I'm not sure how to give advice on the best use of `Optional.empty` here, because I feel like `Optional` is bad at communicating error states and I've felt frustrated with trying to use it here myself, and
- You don't want the result to be a partly-filled data structure. I empathize, but I think a partial response like that would be more consistent with the way the other deserialization methods are written. If I understand correctly, it seems to be working with a best-effort approach, where it'll do its best to return something even if it might have holes in it.

keturn reviewed (changes requested):
Please add a test for the case that was crashing.

keturn commented on subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/coreTypes/GenericMapTypeHandler.java:
```diff
@@ -61,26 +61,45 @@ private PersistedData serializeEntry(Map.Entry<K, V> entry, PersistedDataSeriali
     @Override
     public Optional<Map<K, V>> deserialize(PersistedData data) {
-        if (!data.isArray()) {
+        if (!data.isArray() || data.isValueMap()) {
```
Why the `|| data.isValueMap()` here?

skaldarnar replied:
For some reason the `PersistedData` for a JSON object returns `true` for `isArray()` - and not only for actually JSON arrays as I originally thought. 

A JSON object is also a "value map", so we use this here to identify whether we are looking at the expected format or not.

keturn replied:
:grimacing: is that specific to JSON? is it a bug in PersistedData? 

This should be a straightforward precondition check, but I think the conditional is non-obvious, and that points to something that needs fixing elsewhere (in a separate issue).

jdrueckert replied:
tracked in #5069

skaldarnar commented:
The type handler interface documents that `Optional.empty` corresponds to a failed deserialization. It also has a couple of convenience methods to return `null` instead, or use a default value, or throw an exception `DeserializationException`.

https://github.com/MovingBlocks/Terasology/blob/b367a180d4e32cc0e573f6567f6db54987c092a3/subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/TypeHandler.java#L39-L45

When we talked about this offline, I raised similar concerns and comments to what @keturn mentioned above. Half-parsed components or prefabs are most likely no good to anybody, so being more strict is likely a good choice. To avoid a situation where we change it for one thing but not the others I'd like to take a bit of time to figure out _how_ we actually want this code to behave and look like, and then do a sweep over it to improve all (or most) places where we use type handlers.

In any case, I'd like to be able to assemble errors (and error message) bottom up to be able to present the user with a more meaningful error message on where they have to look to fix broken prefab.

Coming from functional programming in Scala and ReasonML I'm fond of using types to express logic. In this particular case, I'm looking at the [Result](https://en.wikipedia.org/wiki/Result_type) type, c.f. its variant in [Kotlin](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-result/). I think the "more Java" way is to use plain exceptions...
In both cases, we should avoid logging way down in the stack (very similar to [Log and Throw](https://jqassistant.org/log-and-throw/) anti-pattern).

I believe for starters we can try to work with what we already have in the type handler interface, and gradually improve it (with a bit of focus on it for the next week(s)). Let's make this type handler more strict, and also apply it to others. Try to make better use of the information that something went wrong higher hup in the stack.

@keturn @jdrueckert what would be your take on this?

jdrueckert commented:
> Please add a test for the case that was crashing.

done :heavy_check_mark:

jdrueckert commented:
> I believe for starters we can try to work with what we already have in the type handler interface, and gradually improve it (with a bit of focus on it for the next week(s)). Let's make this type handler more strict, and also apply it to others. Try to make better use of the information that something went wrong higher hup in the stack.
> 
> @keturn @jdrueckert what would be your take on this?

Sounds fair to me. I adjusted the existing test case and added some more incl. a test case to test that a half-way valid map is still not half-way returned according to your "let's make this type handler more strict".
I'm also very much in favor to apply it to other type handlers as well as soon as possible.
I don't see it happening across all type handlers before the next playtest, but I'd like to get as far as possible with the capacity I have.

skaldarnar commented on .idea/kotlinc.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="KotlinJpsPluginSettings">
+    <option name="version" value="1.6.21" />
```
Where does this come from? 👀 

(When opening Terasology in IntelliJ on my Mac I got a warning about some old Kotlin version 1.4.x, and that it switched to use 1.6.21. I did not find any references to the specific version it warned about, though. Does this config file pin the version to something reasonable?)

@keturn any idea? IntelliJ help on [How to manage projects under Version Control Systems](https://intellij-support.jetbrains.com/hc/en-us/articles/206544839-How-to-manage-projects-under-Version-Control-Systems) seems to suggest that the file can be shared...

> What needs to be shared:
>
> 1. All files under the .idea directory in the project root except the items that store user-specific settings: ...

keturn replied:
lets separate this out; having it in here is making it more complicated to switch PR branches. #5068

skaldarnar commented on subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/coreTypes/GenericMapTypeHandler.java:
```diff
@@ -61,26 +61,45 @@ private PersistedData serializeEntry(Map.Entry<K, V> entry, PersistedDataSeriali
-        if (!data.isArray()) {
+        if (!data.isArray() || data.isValueMap()) {
+            logger.warn("Incorrect map format detected: object instead of array.\n" +
+                    "Expected\n" +
+                    "  \"mapName\": [\n" +
+                    "    { \"key\": \"...\", \"value\": \"...\" }\n" +
+                    "  ]\n" +
+                    "but found \n'{}'", data);
```
Used in multiple places, maybe create a helper method to be used here and below.

```suggestion
                    getUsageInfo(data));
```

```java
private String getUsageInfo(PersistedData data) {
	return "Expected\n" +
                    "  \"mapName\": [\n" +
                    "    { \"key\": \"...\", \"value\": \"...\" }\n" +
                    "  ]\n" +
                    "but found \n'" + data + "'";
}
```

jdrueckert replied:
done :heavy_check_mark:

skaldarnar commented on subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/coreTypes/GenericMapTypeHandler.java:
```diff
@@ -61,26 +61,45 @@ private PersistedData serializeEntry(Map.Entry<K, V> entry, PersistedDataSeriali
+            PersistedData rawValue = kvEntry.get(VALUE);
+            if (rawKey == null || rawValue == null) {
+                logger.warn("Incorrect map format detected: missing map entry with \"key\" or \"value\" key.\n" +
+                        "Expected\n" +
+                        "  \"mapName\": [\n" +
+                        "    { \"key\": \"...\", \"value\": \"...\" }\n" +
+                        "  ]\n" +
+                        "but found \n'{}'", data);
```
```suggestion
                        getUsageInfo(data));
```

jdrueckert replied:
done :heavy_check_mark:

skaldarnar commented on subsystems/TypeHandlerLibrary/src/test/java/org/terasology/persistence/typeHandling/coreTypes/GenericMapTypeHandlerTest.java:
```diff
@@ -32,6 +31,39 @@ GenericMapTypeHandler.VALUE, new PersistedLong(TEST_VALUE)
             ))
     ));
 
+    private final PersistedData testDataMalformatted = new PersistedValueArray(List.of(
```
Maybe we can add the JSON equivalent we want to express here in a comment? Probably helps in both understanding the test case and figuring out if the data strucutre represents the thing we want to test.

jdrueckert replied:
done :heavy_check_mark:

keturn replied:
:worried: It's also the textbook example of "comment that's out of date as soon as you change the code."

If you think it helps despite that, go ahead and keep it, but I'm a little concerned.

Also note Javadoc does not have triple-quote code blocks. Either change these to JavaDoc syntax, or do not start the comment with the double-asterisk `/**` so IntelliJ and javdoc don't treat it as javadoc.

keturn replied:
I've replaced the triple-quotes.

keturn commented on subsystems/TypeHandlerLibrary/src/test/java/org/terasology/persistence/typeHandling/coreTypes/GenericMapTypeHandlerTest.java:
```diff
@@ -44,23 +132,69 @@ void testDeserialize() {
+    }
+
+    @Test
+    @DisplayName("A map containing both, correctly and incorrectly formatted data, fails deserialization (returns empty `Optional`)")
+    void testDeserializeWithValidAndInvalidEntries() {
+        var th = new GenericMapTypeHandler<>(
+                new StringTypeHandler(),
+                new LongTypeHandler()
+        );
+
+        assertThat(th.deserialize(testDataValidAndInvalidMix)).isEmpty();
```
This is the case I'm a little worried about, since it's a change from the previous behavior. Not that the previous behavior was _better,_ but it's one of those things where something might have been implicitly depending on it.

But this has a test case now, and the hypothetical I'm worrying about doesn't, so this wins.

skaldarnar replied:
I don't think (or hope) something was depending on it. Rather, we'll see some things either work out better or even worse than before, hopefully just epxosing the underlying problem better.

jdrueckert replied:
My expectation is that if this makes something work worse than before, then it's mainly making the underlying issue more visible than before.

> I'm curious: which case was it that you initially bumped in to and prompted this PR?

I ran into the crash after making `ChangingBlocks` use uris instead of strings for the map key.
The prefabs in PlantPack that I tested this were working before with the basic string key type map deserialization but didn't have the format required by the generic map type handler (outer array, entries with "key" and "value" keys). So the three malformatting test cases are covering this now.

keturn reviewed (approved):
Alright, lets go ahead with this.

I appreciate that you've covered a range of cases with the tests, but I'm curious: which case was it that you initially bumped in to and prompted this PR?

skaldarnar commented on subsystems/TypeHandlerLibrary/src/test/java/org/terasology/persistence/typeHandling/coreTypes/GenericMapTypeHandlerTest.java:
```diff
@@ -25,14 +25,102 @@ class GenericMapTypeHandlerTest {
+     * JSON equivalent:
+     * <pre><code>
+     * {
+     *   "health:baseRegen": -1"
```
```suggestion
     *   "health:baseRegen": -1
```

skaldarnar reviewed (approved):
@keturn The issue we found was due to the prefab using the JSON object format to specify the mapping after switching the type to be `SimpleUri`.

```json
{
  "plantpack:radish1": 3000,
  "plantpack:radish2": 2000
}
```

We currently have two variants for defining maps:

- `String` -> `V` if the keys are Strings, value can be anything, handled by a different type handler
  ```json
	  {
	    "plantpack:radish1": 3000,
	    "plantpack:radish2": 2000
	  }  
  ```

- `K` -> `V` where both keys and values can be arbitrary objects that can be de-/serialized
  ```json
	  [
	    { "key": "plantpack:radish1": , "value": 3000 },
	    { "key": "plantpack:radish2": , "value": 2000 },
	  ]  
  ```
  
  In the future, we probabyl want to add a third option (or replace/enhance the first one) if the key type `K` can be parsed from a plain String:
  
- `K` -> `V` if the keys are Strings which can be parsed to some type `K` (e.g., all the URIs and Names), value can be anything, handled by a different type handler
  ```json
	  {
	    "plantpack:radish1": 3000,
	    "plantpack:radish2": 2000
	  }  
  ```
